File: flexinput/src/main/java/com/lytefast/flexinput/adapters/AddContentPagerAdapter.java
package com.lytefast.flexinput.adapters;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.support.annotation.DrawableRes;
import android.support.annotation.StringRes;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.content.res.AppCompatResources;

import com.lytefast.flexinput.R;
import com.lytefast.flexinput.fragment.CameraFragment;
import com.lytefast.flexinput.fragment.FilesFragment;
import com.lytefast.flexinput.fragment.PhotosFragment;


/**
 * {@link FragmentPagerAdapter} for the "Add Content" {@link android.support.v4.view.ViewPager}
 * allowing the user to set custom pages via a list of {@link PageSupplier}s.
 *
 * @author Sam Shih
 * @see #createDefaultPages() for the default set of pages that are supported
 */
public class AddContentPagerAdapter extends FragmentPagerAdapter {

  private final PageSupplier[] pageSuppliers;


  public AddContentPagerAdapter(final FragmentManager childFragmentManager, PageSupplier... pageSuppliers) {
    super(childFragmentManager);
    this.pageSuppliers =
        (pageSuppliers == null || pageSuppliers.length == 0) ? createDefaultPages() : pageSuppliers;
  }

  @Override
  public Fragment getItem(final int position) {
    return pageSuppliers[position].createFragment();
  }

  @Override
  public int getCount() {
    return pageSuppliers.length;
  }

  public void initTabs(final Context context, final TabLayout tabLayout) {
    final ColorStateList iconColors =
        AppCompatResources.getColorStateList(context, R.color.tab_color_selector);

    // Color the existing tabs
    for (int i = 0; i < tabLayout.getTabCount(); i++) {
      setIconColor(iconColors, tabLayout.getTabAt(i));
    }

    // Create tabs for the adapter pages
    for (PageSupplier page : pageSuppliers) {
      TabLayout.Tab tab = tabLayout.newTab()
          .setIcon(page.icon)
          .setContentDescription(page.contentDesc);
      setIconColor(iconColors, tab);
      tabLayout.addTab(tab);
    }
  }

  private void setIconColor(final ColorStateList iconColors, final TabLayout.Tab tab) {
    Drawable icon = tab.getIcon();
    if (icon != null) {
      icon = DrawableCompat.wrap(icon);
      DrawableCompat.setTintList(icon, iconColors);
    }
  }

  public static abstract class PageSupplier {
    @DrawableRes protected final int icon;
    @StringRes protected final int contentDesc;

    public PageSupplier(@DrawableRes int icon, @StringRes final int contentDesc) {
      this.icon = icon;
      this.contentDesc = contentDesc;
    }

    public abstract Fragment createFragment();
  }

  public static PageSupplier[] createDefaultPages() {
    return new PageSupplier[]{
        new PageSupplier(R.drawable.ic_file_24dp, R.string.attachment_files) {
          @Override
          public Fragment createFragment() {
            return new FilesFragment();
          }
        },
        new PageSupplier(R.drawable.ic_image_24dp, R.string.attachment_photos) {
          @Override
          public Fragment createFragment() {
            return new PhotosFragment();
          }
        },
        new PageSupplier(R.drawable.ic_add_a_photo_24dp, R.string.attachment_camera) {
          @Override
          public Fragment createFragment() {
            return new CameraFragment();
          }
        }
    };
  }
}

File: flexinput/src/main/java/com/lytefast/flexinput/adapters/AddContentPagerAdapter.kt
package com.lytefast.flexinput.adapters

import android.content.Context
import android.content.res.ColorStateList
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.graphics.drawable.DrawableCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import com.google.android.material.tabs.TabLayout
import com.lytefast.flexinput.R
import com.lytefast.flexinput.adapters.AddContentPagerAdapter.PageSupplier
import com.lytefast.flexinput.fragment.CameraFragment
import com.lytefast.flexinput.fragment.FilesFragment
import com.lytefast.flexinput.fragment.PhotosFragment


/**
 * [FragmentPagerAdapter] for the "Add Content" [androidx.viewpager.widget.ViewPager]
 * allowing the user to set custom pages via a list of [PageSupplier]s.
 *
 * @author Sam Shih
 * @see .createDefaultPages
 */
class AddContentPagerAdapter(childFragmentManager: FragmentManager,
                             private vararg val pageSuppliers: PageSupplier)
  : FragmentPagerAdapter(childFragmentManager) {

  override fun getItem(position: Int): Fragment = pageSuppliers[position].createFragment()

  override fun getCount(): Int = pageSuppliers.size

  fun initTabs(context: Context, tabLayout: TabLayout) {
    val iconColors = AppCompatResources.getColorStateList(context, R.color.tab_color_selector)

    // Color the existing tabs
    for (i in 0 until tabLayout.tabCount) {
      tabLayout.getTabAt(i)?.setIconColor(iconColors)
    }

    // Create tabs for the adapter pages
    pageSuppliers
        .map {
          tabLayout.newTab()
              .setIcon(it.icon)
              .setIconColor(iconColors)
              .setContentDescription(it.contentDesc)
        }
        .forEach { tabLayout.addTab(it) }
  }

  private fun TabLayout.Tab.setIconColor(iconColors: ColorStateList) : TabLayout.Tab {
    this.icon?.let {
      val icon = DrawableCompat.wrap(it)
      DrawableCompat.setTintList(icon, iconColors)
      this.icon = icon
    }
    return this
  }

  abstract class PageSupplier(@DrawableRes val icon: Int, @StringRes val contentDesc: Int) {

    abstract fun createFragment(): Fragment
  }

  companion object {

    fun createDefaultPages(): Array<PageSupplier> {
      return arrayOf(
          object : PageSupplier(R.drawable.ic_image_24dp, R.string.attachment_photos) {
            override fun createFragment(): Fragment = PhotosFragment()
          },
          object : PageSupplier(R.drawable.ic_file_24dp, R.string.attachment_files) {
            override fun createFragment(): Fragment = FilesFragment()
          },
          object : PageSupplier(R.drawable.ic_add_a_photo_24dp, R.string.attachment_camera) {
            override fun createFragment(): Fragment = CameraFragment()
          })
    }
  }
}
